api/core/storage/file_storage.py

```diff
@@ -0,0 +1,15 @@
+from typing import NamedTuple, Protocol
+
+
+class FileData(NamedTuple):
+    contents: bytes
+    content_type: str | None = None
+    filename: str | None = None
+
+
+class CouldNotStoreFileError(Exception):
+    pass
+
+
+class FileStorage(Protocol):
+    async def store_file(self, file: FileData, folder_path: str) -> str: ...
```

Comment:
just curious why do you use a Protocol this time and not an abstract class ?

Reply:
I think it's safe to use `Protocol` instead of ABC most times. The only upside of ABC is that you can add default implementation for functions (and inference at run time) but here there is no point and I only care about Protocol conformance.

`Protocol` has the benefit of not needing explicit subclassing which can be very useful.